[Feat] 상세화면의 컬렉션 뷰 셀 및 헤더를 구현합니다.

MinwooJe opened the pull request:
# 배경
#78 
<img width="394" alt="스크린샷 2025-01-09 오후 6 26 50" src="https://github.com/user-attachments/assets/5e5735ff-83b3-4234-9607-a7604ffbe591" />  


# 작업 내용
## 1. 팝업 상세화면 구조 변경
### 배경
기존에는 스크롤뷰로 구현하려 하였으나, 후기 UI를 구현하기 위해 테이블 뷰 혹은 컬렉션 뷰를 고려하게 되었습니다.

스크롤 뷰를 유지하면 스크롤 뷰 위에 컬렉션 뷰(후기 UI)를 올리는 형태가 되는데, 이는 다음과 같은 문제를 유발할 수 있습니다.
- 콘텐츠 크기 계산 및 스크롤 동기화 문제
- 중첩된 스크롤 뷰로 인해 사용성 저하

이는 메인화면 구현 시 발생한 문제와 동일한 문제이므로 메인화면처럼 하나의 컬렉션 뷰로 전환하기로 결정하였습니다.
> **참고**
> #73 - [feat] 메인화면 캐러셀뷰의 스크롤을 구현합니다. 
> 2. MainCarouselView를 mainCollectionView의 첫 번째 섹션의 헤더로 추가

### 선택
위 문제를 해결하기 위해 스크롤 뷰를 컬렉션 뷰로 전환하기로 결정하였습니다.  
테이블 뷰가 아닌 컬렉션 뷰를 선택한 이유는 다음과 같습니다.
- 섹션마다 독립적인 헤더를 설정 가능
- 섹션마다 다른 레이아웃을 설정 가능


### 설계
컬렉션 뷰를 기반으로 다음과 같은 구조를 설계하였습니다.
<img src="https://github.com/user-attachments/assets/611599ec-36be-4b4c-840f-ddb388044ef5" width=50%>

컬렉션뷰의 레이아웃은 `UICollectionViewCompositionalLayout`을 이용하였습니다.  

그 이유는 각 섹션에서 데이터의 길이(제목, 해시태그 개수, 팝업스토어 정보 텍스트의 길이, 후기 텍스트의 길이 등)에 따라 셀의 높이가 동적으로 변경됩니다.  

이때, `Flow Layout`은 `UICollectionViewDelegateFlowLayout`의 `sizeForItemAt` 메서드에서 별도로 높이를 계산해야 하므로 코드가 복잡해집니다.  

그러나 `Compostional Layout`은 `.estimated` 설정을 통해 셀의 동적 높이를 간단하게 구현할 수 있기 때문입니다.

## 2. `MainCarouselViewModelProtocol` 수정
### 프로퍼티 타입, 메서드 반환 타입 변경
`MainCarouselViewModelProtocol`은 캐러셀 뷰의 데이터 소스 역할을 하는 뷰모델이 채택하는 프로토콜입니다.  
캐러셀 뷰의 재사용과 각 뷰모델의 다형성을 위해 구현하였습니다.   

그러나 기존에는 아래와 같이 메인화면에 종속되는 `PopupPreview` 타입을 가지고 있었습니다.
```swift
// 변경 전
protocol MainCarouselViewModelProtocol {
    var carouselPopupImage: [PopupPreview] { get set }
 
    var carouselImagePublisher: (() -> Void)? { get set }

    func numbersOfCarouselImage() -> Int

    func provideCarouselImage(at index: Int) -> PopupPreviewData?
}
```
따라서 상세 화면에서 사용하기엔 적합하지 않았습니다.  

캐러셀 뷰의 데이터는 이미지이므로 `PopupPreview` 타입에서 `UIImage` 타입으로 변경하였습니다.
```swift
// 변경 후
import UIKit

protocol MainCarouselViewModelProtocol: AnyObject {
    var carouselPopupImage: [UIImage] { get set }

    var carouselImagePublisher: (() -> Void)? { get set }

    func numbersOfCarouselImage() -> Int

    func provideCarouselImage() -> [UIImage]
}
```
또한 기존의 `provideCarouselImage()` 메서드의 파라미터로 `index`를 받고있었는데, 이는 `MainSceneViewModel.providePopupPreviewData(of:, at:)` 메서드에 종속적인 설계였습니다.   

추상 타입이 구체 타입에 의존하는 형태였고, 따라서 상세화면의 뷰 모델에서 사용할 수 없을 것이라 판단하여 해당 파라미터를 삭제하였습니다. 

### `AnyObject`를 준수
> **레퍼런스**
> https://babbab2.tistory.com/118
> https://medium.com/@b9d9/anyobject-를-준수하는-protocol-의-몰랐던-사실-50b5052859a1  

```swift
final class MainCarouselView: UIView {
    private var viewModel: MainCarouselViewModelProtocol?
    ...
    private func bind(to viewModel: MainCarouselViewModelProtocol) {
        viewModel.carouselImagePublisher = { [weak self] in
            guard let self else { return }
            self.imagePageControl.numberOfPages = viewModel.numbersOfCarouselImage()
            self.carouselCollectionView.reloadData()
        }
    }
}
```
`AnyObject`를 준수하지 않았을 때, 위 코드의 `bind(to:)`메서드의 `viewModel.carouselImagePublisher = { }` 부분에서  
`Cannot assign to property viewModel is let constant`라는 에러가 발생하였습니다.  

이유는 다음과 같습니다:
1. **프로토콜 타입의 불분명한 값 타입/참조 타입**  
   `MainCarouselViewModelProtocol`이 `AnyObject`를 준수하지 않는다면, Swift는 해당 프로토콜을 **값 타입**(구조체, 열거형)에서도 사용할 수 있다고 판단합니다.  
   이로 인해, `bind(to:)` 메서드의 파라미터로 전달된 `viewModel`이 값 타입일 가능성을 고려하게 됩니다.
2. **파라미터는 기본적으로 let 상수로 취급**  
   Swift에서는 메서드 파라미터가 기본적으로 `let` 상수로 선언되므로 `bind(to:)` 메서드의 `viewModel` 또한 불변입니다.  
   값 타입인 경우, 상수로 선언되었을 때 내부의 `var` 프로퍼티조차 변경이 불가능합니다.  
   따라서 `carouselImagePublisher`를 설정하려는 시도에서 컴파일 에러가 발생한 것입니다.
3. **해결책: AnyObject를 준수하도록 제한**  
   `MainCarouselViewModelProtocol`이 `AnyObject`를 준수하도록 변경하면, 해당 프로토콜은 **클래스 타입**에서만 채택할 수 있습니다.  
   클래스는 참조 타입이므로 `let`으로 선언되어도, 내부의 `var` 프로퍼티를 수정할 수 있습니다.

따라서 `MainCarouselViewModelProtocol`이 `AnyObject`를 준수하여 해당 프로토콜을 따르는 타입이 클래스 타입임을 보장시키고, 이를 통해 컴파일 에러를 해결할 수 있었습니다.

## 3. `PopupInfoReviewSegmentHeaderViewDelegate` 구현
컬렉션 뷰의 헤더로서 `PopupInfoReviewSegmentHeaderView`를 구현하였습니다.  
상세화면 컬렉션 뷰의 설계에서는 `segmentedControl.selectedSegmentIndex`가 변경될 때 마다 컬렉션 뷰의 섹션을 변경해줘야 합니다.  
따라서 `PopupInfoReviewSegmentHeaderView`의 `selectedSegmentIndex`를 상위 뷰로 전달하기 위해 델리게이트를 구현하였습니다.

## 4. UI 구현 및 수정
상세 화면 컬렉션 뷰에 필요한 셀과 헤더(`ReusableView`)를 구현하였습니다.  

# 테스트 방법
아직 컬렉션 뷰를 구현하지 않아 테스트 할 수 없습니다ㅠㅠ  
혹시 보고싶은 뷰가 있다면 `ViewController.swift`에서 초기화 하면 볼 수 있습니다!

# 리뷰노트
작업 도중 설계가 변경되고, 수정 사항이 많아 커밋 히스토리가 지저분합니다ㅠ
`feat: 세그먼트 컨트롤 구현, 0b72442d1a115521b8ddec09cce0c325394a0a0b`부터 `feat: 팝업스토어 상세화면 정보 구현, 0b7b926d8f61ce667aeab5cdff958ba35bc25f03` 까지의 첫 세 커밋은 넘어가주셔도 됩니다!

`Domain/Model/Entities/PopupDetail.swift` 파일은 변경 예정이므로 넘어가주세요!

MinwooJe committed:
feat: 세그먼트 컨트롤 구현

MinwooJe committed:
design: Location, Link 이미지 수정 및 추가

MinwooJe committed:
feat: 팝업스토어 상세화면 정보 구현

MinwooJe committed:
rename: Presentation/Main/View를 메인화면과 상세화면으로 나누어 폴더링

MinwooJe committed:
refactor: MainDetailViewController에서 PopupTitleCollectionViewCell 관련 코드 이동

MinwooJe committed:
refactor: MainDetailViewController에서 세그먼트 컨트롤을 PopupInfoDescriptionCollectionReusableView로 분리

MinwooJe committed:
refactor: PopupDetailInfoView의 타입을 UICollectionViewCell로 변경

MinwooJe committed:
refactor: MainCarouselView를 담는 CarouselCollectionReusableView 구현

MinwooJe committed:
refactor: 스크롤 뷰 관련 코드 삭제

MinwooJe committed:
feat: PopupDetail 엔티티 정의

MinwooJe committed:
feat: 후기 별점 바 구현

MinwooJe committed:
design: 별점 이미지 추가 및 수정

MinwooJe committed:
feat: 별점 뷰 구현

- 생성자의 isUserInteractionEnabled가 true이면 리뷰 쓰기 화면에서의 드래그로 별점 조절 기능 가능
false면 후기 탭처럼 읽기 전용으로만 동작

MinwooJe committed:
feat: 후기 점수 셀 구현

MinwooJe committed:
chore: 개행 수정

MinwooJe committed:
design: 후기 정렬 버튼 이미지 추가

MinwooJe committed:
feat: 리뷰 정렬 헤더 구현

MinwooJe committed:
feat: 리뷰 컬렉션뷰 셀 구현

MinwooJe committed:
fix: 별점 뷰 configureRating를 이용해 점수 변경 시 별점 이미지 갱신안되는 오류 수정

MinwooJe committed:
rename: ~CollectionReusableView -> HeaderView로 접미사 변경

MinwooJe committed:
refactor: MainCarouselViewModelProtocol 수정

- carouselPopupImage 타입, provideCarouselImage() 반환 타입 변경
`PopupPreview`를 반환하면 안되고, `UIImage` 타입을 반환해야 하기 때문

MinwooJe committed:
feat: PopupDetailViewModel 선언 및 MainCarouselViewModelProtocol 채택

MinwooJe committed:
feat: 컬렉션 뷰 셀, 헤더에 Public Interface 추가 및 수정

MinwooJe committed:
refactor: 레이아웃 수정

MinwooJe committed:
chore: 오타(Priod -> Period) 및 개행 수정

MinwooJe committed:
feat: PopupInfoReviewSegmentHeaderViewDelegate 구현

segmentedControl.selectedSegmentIndex를 상위 뷰에 전달하기 위함

MinwooJe committed:
Merge branch 'feat/상세화면-UI-구현' into develop

MinwooJe committed:
chore: PopupDetailViewModel에서 UIKit import

snughnu commented on Popcorn-iOS/Source/Presentation/Main/PopupDetailView/RatingDistributionView.swift:
```diff
@@ -0,0 +1,95 @@
+        return label
+    }()
+
+    private lazy var ratingProgressBar: UIProgressView = {
```
저는 뷰 2개를 쓰려고 했을거같은데 `UIProgressView` 는 처음봤어요!
다음에 저도 써보도록 하겠습니다~

MinwooJe replied:
진행정도를 나타내는 UI는 프로그레스뷰 사용하시면 편할거에요!

snughnu commented on Popcorn-iOS/Source/Presentation/Main/PopupDetailView/RatingDistributionView.swift:
```diff
@@ -0,0 +1,95 @@
+        ratingCountLabel.text = String(ratingCount)
+    }
+
+    func higlightMaximumDistribution(text: ReviewDistribution) {
```
highlight에서 h가 빠진거같아요!

MinwooJe replied:
앗 감사합니다!!

MinwooJe replied:
관련 코드 수정사항이 있어 같이 반영하겠습니다~

snughnu commented on Popcorn-iOS/Source/Presentation/Main/PopupDetailView/RatingDistributionView.swift:
```diff
@@ -0,0 +1,95 @@
+    }
+}
+
+// MARK: - Public Interface
```
public만 따로 묶는 이유가 있을까요?
만약 기능별로 정리하다가 public과 private가 섞이면 어떻게 하는게 좋을까요?

MinwooJe replied:
Public만을 따로 묶는다기 보다는, 외부 클래스와 소통하는 Public Interface 메서드만 따로 묶는것입니다!

Public과 Private이 섞인다는 말이 이해가 잘 가지는 않지만, Public Interface는 세부적인 구현과 private 멤버들은 숨기고, 소통할 수 있는 API라고 생각하시면 됩니다. 자동차의 동작원리를 몰라도 핸들과 페달로만 운전할 수 있는 것 처럼요!

개인적으로 한 군데에 묶어두면 유지보수 및 수정 시 편해서 이렇게 사용하고 있습니다.

snughnu replied:
넵 알겠습니다~

snughnu commented on Popcorn-iOS/Source/Presentation/Main/PopupDetailView/ReusableView/PopupDetailInfoCollectionViewCell.swift:
```diff
@@ -0,0 +1,229 @@
+    private let locationLabel: UILabel = {
+        let label = UILabel()
+        label.popcornMedium(text: "주소", size: 15)
+        label.numberOfLines = 0
```
개인적인 생각으로는 주소는 한줄로 표시되는게 좋을거 같아요
아래 코드의 StackView로 잡을 때 fillEqually로 설정한 부분에서 UI가 조금 이상해질 거 같습니다!

MinwooJe replied:
흠 저는 UI의 일관성을 주기 위해 두 줄 로 표시되더라도 폰트 크기를 맞추는 것이 좋을 것 같다라고 생각했습니다.
폰트 크기가 작아져도 한 줄로 맞춘다면 위치의 일관성...?이 있어서 장단점이 있겠네요!

디자이너님께 한번 여쭤보겠습니다!

MinwooJe replied:
스택뷰의 `fillEqaully`는 저도 미처 생각하지 못한 부분이네요 감사합니다!!!

MinwooJe replied:
두 줄로 표시하는 것으로 결정됐습니다!  
아래는 관련 피그마 코멘트입니다~
https://www.figma.com/design/DVerEwsns12Js5WtTijImy?node-id=62-504&m=dev#1086685929

snughnu commented on Popcorn-iOS/Source/Presentation/Main/PopupDetailView/ReusableView/PopupDetailInfoCollectionViewCell.swift:
```diff
@@ -0,0 +1,229 @@
+        return imageView
+    }()
+
+    private let officialLinkLabel: UILabel = {
```
저는 링크가 버튼이라고 생각했는데 레이블에 링크를 달 수 있나보군요
다음에 저도 해보겠습니다!

MinwooJe replied:
엇 저도 그거까지는 생각하진 못했네요...ㅎㅎ
관련 내용 찾아보고 수정사항 생기면 반영하겠습니다!

MinwooJe replied:
> **레퍼런스**
> https://jellysong.tistory.com/127

`UILabel` 대신 `UITextView`를 사용하는 것이 나을 것 같습니다!
`UITextView`에는 dataDetectorTypes 속성이 있어, phone numbers, http links 등을 자동으로 인식해 URL 로 변환해주고 필요한 기능을 수행할 수 있네요!

`UITextView`로 변경 후 다시 푸쉬하겠습니다~

snughnu replied:
`UITextView`도 있군요!

snughnu commented on Popcorn-iOS/Source/Presentation/Main/PopupDetailView/ReusableView/PopupDetailInfoCollectionViewCell.swift:
```diff
@@ -0,0 +1,229 @@
+        stackView.axis = .vertical
+        stackView.spacing = 4
+        stackView.alignment = .leading
+        stackView.distribution = .fillEqually
```
위에서 주소가 한 줄이 좋다고 생각한 부분에 해당하는 코드입니다!
`fillEqually`로 설정하고 `locationLabel`이 여러줄이 되면 레이아웃이 이상하게 잡힐거 같은데 괜찮았나요?

MinwooJe replied:
`.fillProportionally`로 변경했습니다!
감사합니다!!!!

snughnu commented on Popcorn-iOS/Source/Presentation/Main/PopupDetailView/ReusableView/PopupDetailInfoCollectionViewCell.swift:
```diff
@@ -0,0 +1,229 @@
+    private func configureLayout() {
+        let reservationStartChatButtonStackViewTopConstraint = reservationStartChatButtonStackView.topAnchor
+            .constraint(equalTo: popupInfoStackView.bottomAnchor, constant: 30)
+        reservationStartChatButtonStackViewTopConstraint.priority = .defaultLow
```
우선순위에 대한 코드가 어떤 부분에서 사용되고 있는지 잘 모르겠습니다!

MinwooJe replied:
현재 코드에서는 사용되지 않는데.. 깜빡하고 안지운 것 같애요ㅠㅠ
삭제해서 다시 푸쉬하겠습니다!

snughnu commented on Popcorn-iOS/Source/Presentation/Main/PopupDetailView/ReusableView/PopupInfoReviewSegmentHeaderView.swift:
```diff
@@ -0,0 +1,126 @@
+        let view = UIView()
+        view.backgroundColor = .black
+        view.translatesAutoresizingMaskIntoConstraints = false
+        view.heightAnchor.constraint(equalToConstant: 3).isActive = true
```
이 부분은 Configure UI 에 있는게 좋지 않을까요?

MinwooJe replied:
```swift
        view.translatesAutoresizingMaskIntoConstraints = false
        view.heightAnchor.constraint(equalToConstant: 3).isActive = true
```
이 부분 말씀하시는거 맞을까요?
`view.translatesAutoresizingMaskIntoConstraints = false`는 중복으로 들어가있어서 빼도 될 것 같습니다!

MinwooJe replied:
저도 원래는  `configureLayout()` 메서드에 넣었는데 separatorView의 width, height 같은 경우에는 초기화 시 지정하였습니다.

그 이유는 아래와 같습니다.
- `separatorView`의 width, height는 특히 자기 자신만의 속성...?이라고 생각함.
- `configureLayout()` 메서드를 간결하게 작성하기 위함.

`configureLayout()` 메서드에 작성한다면, 레이아웃 관련 코드를 한 데 모아 관리할 수 있다는 장점이 있을 것 같아요!

사실 저도 이렇게 작성하면서 고민은 많았는데, 컨벤션을 정하면 좋을 것 같습니다!

저는 레이아웃 관련 코드를 `configureLayout()` 메서드에 모았으니, `configureLayout()`메서드로 옮겨도 좋을 것 같다고 생각됩니다!

성훈님은 어떻게 생각하시나요?

snughnu replied:
저도 로그인화면과 회원가입화면에서 민우님과 똑같이 생각하여 코드를 작성했었습니다.
PR에서 잠시 얘기했었던 기억이 있는데 `configureLayout`에 넣자고 결론이 나왔는데 구두로 했던거같습니다. 기록을 못찾겠네요 ^^;
지금 다시 정해볼까요?

MinwooJe replied:
아하 저도 깜빡했네요ㅠㅠ
`configureLayout`에 넣는걸로 정해요!

snughnu commented on Popcorn-iOS/Source/Presentation/Main/PopupDetailView/ReusableView/PopupInfoReviewSegmentHeaderView.swift:
```diff
@@ -0,0 +1,126 @@
+        let view = UIView()
+        view.backgroundColor = UIColor(resource: .popcornGray2)
+        view.translatesAutoresizingMaskIntoConstraints = false
+        view.heightAnchor.constraint(equalToConstant: 3).isActive = true
```
이 부분도 마찬가지로 Configure UI 에 있는게 좋지 않을까요?

MinwooJe replied:
위 답변과 같은 이유입니다!

snughnu commented on Popcorn-iOS/Source/Presentation/Main/PopupDetailView/ReusableView/PopupRatingCollectionViewCell.swift:
```diff
@@ -0,0 +1,142 @@
+    private let separatorView: UIView = {
+        let view = UIView()
+        view.backgroundColor = UIColor(resource: .popcornGray2)
+        view.widthAnchor.constraint(equalToConstant: 1).isActive = true
```
이 부분도 Configure UI에 있는게 좋을거같아요

MinwooJe replied:
위 답변과 같은 이유입니다!

snughnu commented on Popcorn-iOS/Source/Presentation/Main/PopupDetailView/ReusableView/ReviewCollectionViewCell.swift:
```diff
@@ -0,0 +1,201 @@
+        let layout = UICollectionViewFlowLayout()
+        layout.scrollDirection = .horizontal
+        layout.minimumLineSpacing = 6
+        layout.itemSize = CGSize(width: 60, height: 60)
```
여기서 고정된 크기를 사용해도 괜찮나요?
저는 고정된 상수를 사용해도 되는 부분과 비율에 맞게 작성해야하는 부분이 헷갈리는거 같습니다. ㅠㅠ

MinwooJe replied:
저도 사실 고정된 상수를 사용해도 되는 부분과 비율과 맞게 작성해야하는 부분을 명확히 구분하는 기준은 아직 찾지 못했습니다..ㅠ

이 코드에서 6이라는 상수를 사용한 이유는 컬렉션 뷰 컨텐츠의 가로 크기는 제한이 없기 때문입니다.(가로 스크롤 컬렉션 뷰이기 때문)

snughnu commented on Popcorn-iOS/Source/Presentation/Main/PopupDetailView/ReusableView/ReviewCollectionViewCell.swift:
```diff
@@ -0,0 +1,201 @@
+        let imageView = UIImageView()
+        imageView.contentMode = .scaleToFill
+        imageView.clipsToBounds = true
+        imageView.image = UIImage(resource: .grayCircle)
```
프로필 이미지 캐릭터 에셋 추가됐으니 그거 사용해도 괜찮을거같아요!

MinwooJe replied:
감사합니다!!
여기서는 placeholder 역할을 하는 이미지를 넣기 위해 `.grayCircle` 이미지를 사용했습니다!
서버에서 데이터 로딩 시간이 길어진다면, 프로필 이미지 캐릭터를 사용하는 것 보다 `.grayCircle`이 더 적절한 것 같더라구요!

snughnu commented on Popcorn-iOS/Source/Presentation/Main/PopupDetailView/ReusableView/ReviewCollectionViewCell.swift:
```diff
@@ -0,0 +1,201 @@
+    private let separatorView: UIView = {
+        let view = UIView()
+        view.backgroundColor = UIColor(resource: .popcornGray2)
+        view.widthAnchor.constraint(equalToConstant: 1).isActive = true
```
이 부분도 Configure UI에 있는게 좋을거같아요

MinwooJe replied:
위와 같은 이유입니다!

snughnu commented on Popcorn-iOS/Source/Presentation/Main/PopupDetailView/ReusableView/ReviewSortButtonHeaderView.swift:
```diff
@@ -0,0 +1,139 @@
+    private let separatorView: UIView = {
+        let view = UIView()
+        view.backgroundColor = UIColor(resource: .popcornGray4)
+        view.heightAnchor.constraint(equalToConstant: 9).isActive = true
```
이 부분도 Configure UI에 있는게 좋을거같아요

MinwooJe replied:
위와 같은 이유입니다! 
하나하나 확인해주셔서 감사합니다ㅠ

snughnu commented on Popcorn-iOS/Source/Presentation/Main/PopupDetailView/ReusableView/ReviewSortButtonHeaderView.swift:
```diff
@@ -0,0 +1,139 @@
+        return stackView
+    }()
+
+    private let bottomBorder = CALayer()
```
여기서 CALayer가 하는일이 뭔지 잘 모르겠어요!

MinwooJe replied:
![image](https://github.com/user-attachments/assets/359e2bc2-ead3-45e3-af9e-ffc8513d94e6)

이 UI를 구현하기 위해 bottomBorder를 사용했습니다.
저도 지금 보니 `CALayer`를 사용한 이유가 기억나질 않네요...

`CALayer`를 사용하지 않고, `UIView`를 사용해도 동일한 UI를 구현할 수 있더라구요!
굳이 저수준 API인 `CALayer`를 사용하지 않고, UIView를 사용해도 될 것 같습니다!

변경 후 푸쉬하겠습니다~

snughnu reviewed (commented):
파일이 많아질 수 밖에 없는 구조네요 수고하셨습니다..

MinwooJe committed:
chore: 불필요 코드 삭제

코드 수정하며 삭제하지 않은 코드 삭제

MinwooJe committed:
chore: 오타 수정

MinwooJe committed:
refactor: 최빈 평점 종류를 강조하는 로직 수정

다음과 같은 구조로 변경
뷰 모델에서 최빈 평점 종류의 인덱스를 받고
-> cell.configureContents 메서드의 파라미터로 `maximumIndex` 전달
-> cell.configureContents에서는 해당 RatingDistributionView의 `highlightMaximumDistribution` 메서드 호출

hightlightMaxiumDistribution의 파라미터로 "매우 만족", "만족"과 같은 텍스트를 받는다면, 휴먼 에러 발생 가능성이 있기 때문에 변경함.

MinwooJe committed:
design: 스택뷰 spacing 수정

MinwooJe committed:
refactor: bottomborder를 CALayer -> UIView로 변경

MinwooJe committed:
refactor: URL로 나타내기 위해 officialLinkLabel을 TextView로 변경

MinwooJe committed:
chore: UIView 선언 시 heightAnchor를 지정했던 코드 configureLayout() 메서드로 이동
